### PR TITLE
Updated docs

### PR DESCRIPTION
**Pull request by dtrawins**

### CODE REVIEW

File: docs/FaceDetectionScriptExample.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+#### Run the client:
+```bash
+cd example_client
+virtualenv .venv
+. .venv/bin/activate
+pip install -r client_requirements.txt
+mkdir results
+
+python face_detection.py --batch_size 1 --width 300 --height 300
```
[the remark this answers is not in the document]
Comment (dtrawins) [reply docs/FaceDetectionScriptExample.md:58]: those are default values

File: docs/FaceDetectionScriptExample.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+curl --create-dirs https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/1/person-vehicle-bike-detection-crossroad-0078/FP32/person-vehicle-bike-detection-crossroad-0078.bin https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/1/person-vehicle-bike-detection-crossroad-0078/FP32/person-vehicle-bike-detection-crossroad-0078.xml -o model/1/person-vehicle-bike-detection-crossroad-0078.bin -o model/1/person-vehicle-bike-detection-crossroad-0078.xml
+```
+
+### Prepare Model Repository
+
+Refer this link to place your downloaded or converted in a repository to mount it to the OpenVINO&trade; Model Server- [Preparation of Models](./models_repository.md)
+
+### Run the OpenVINO&trade; Model Server Docker Container
+
+Run the OpenVINO&trade; Model Server with the downloaded models. Adjust the shape and batch_size parameter according to the requirement of model. Refer this guide to know more about the parameters.
```
[the remark this answers is not in the document]
Comment (dtrawins) [reply docs/FaceDetectionScriptExample.md:92]: link missing

**Review by mzegla — COMMENTED** [review]
We have some docs in snake_case and others in PascalCase , i.e. **developer_guide.md** and **ModelServerGRPCAPI.md**

Can we unify it and rename some of the documents so they follow the same convention?

File: docs/ModelServerGRPCAPI.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
+# OpenVINO&trade; Model Server gRPC API Documentation
+
+## Introduction 
+This documents gives information about OpenVINO&trade; Model Server gRPC API. It is documented in the proto buffer files in [tensorflow_serving_api](https://github.com/tensorflow/serving/tree/r1.14/tensorflow_serving/apis).
```
Comment (pgierasi) [inline docs/ModelServerGRPCAPI.md:4]: Do we want ot refer to 1.14 TensorFlow version?

File: example_client/README.md [hunk 4]
```diff
@@ -20,43 +111,28 @@ usage: grpc_serving_client.py [-h] --images_numpy_path IMAGES_NUMPY_PATH
...
+| --grpc_address GRPC_ADDRESS | Specify url to grpc service. Default:localhost | 
+| --grpc_port GRPC_PORT | Specify port to grpc service. Default: 9000 |
+| --input_name | Specify input tensor name. Default: input |
+| --output_name | Specify output name. Default: resnet_v1_50/predictions/Reshape_1 |
+| --transpose_input {False,True}|  Set to False to skip NHWC>NCHW or NCHW>NHWC input transposing. Default: True|
+| --transpose_method {nchw2nhwc,nhwc2nchw} | How the input transposition should be executed: nhwc2nchw or nhwc2nchw |
+| --iterations | Number of requests iterations, as default use number of images in numpy memmap. Default: 0 (consume all frames)|
+| --batchsize | Number of images in a single request. Default: 1 |
+| --model_name | Define model name, must be same as is in service. Default: resnet|
+
```
Comment (pgierasi) [inline example_client/README.md:132]: No pipeline_name listed

File: example_client/README.md [hunk 5]
```diff
@@ -1,13 +1,104 @@
...
+* <a href="#rest-api">REST API Client Examples  </a>
 
-## Requirements
+## Requirement
+
+Install client dependencies using the command :
+```
+pip3 install -r client_requirements.txt
+```
+
```
Comment (pgierasi) [inline example_client/README.md:15]: For all of this examples we should specify cmd for starting OMVS and downloading proper models

File: example_client/README.md [hunk 6]
```diff
@@ -20,43 +111,28 @@ usage: grpc_serving_client.py [-h] --images_numpy_path IMAGES_NUMPY_PATH
...
+| --transpose_method {nchw2nhwc,nhwc2nchw} | How the input transposition should be executed: nhwc2nchw or nhwc2nchw |
+| --iterations | Number of requests iterations, as default use number of images in numpy memmap. Default: 0 (consume all frames)|
+| --batchsize | Number of images in a single request. Default: 1 |
+| --model_name | Define model name, must be same as is in service. Default: resnet|
+
+
+- Usage example
+
 ```bash
 python grpc_serving_client.py --grpc_port 9001 --images_numpy_path imgs.npy --input_name data --output_name prob --transpose_input False --labels_numpy lbs.npy
```
Comment (pgierasi) [inline example_client/README.md:137]: --labels_numpy_path

File: example_client/README.md [hunk 7]
```diff
@@ -177,60 +252,113 @@ Detected: 340  Should be: 340
...
+
+- Optional Arguments
+
+| Argument      | Description |
+| :---        |    :----   |
+| -h,--help       | Show help message and exit       |
+| -n NETWORK_NAME, --network_name NETWORK_NAME   |   Network name      |
+| -l INPUT_LAYER, --input_layer INPUT_LAYER | Input layer name |
+| -o OUTPUT_LAYER, --output_layer OUTPUT_LAYER | Output layer name | 
+| -d INPUT_DIMENSION, --input_dimension INPUT_DIMENSION | Input image dimension |
```
Comment (pgierasi) [inline example_client/README.md:282]: --frame_size instead of --input_dimension

File: docs/ovms_quickstart.md [hunk 8]
```diff
@@ -49,18 +49,18 @@ make docker_build
...
+curl --create-dirs https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/1/face-detection-retail-0004/FP32/face-detection-retail-0004.xml https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/1/face-detection-retail-0004/FP32/face-detection-retail-0004.bin -o model/1/face-detection-retail-0004.xml -o model/1/face-detection-retail-0004.bin
 ```
 
 ### Step 4: Start the Model Server Container
 
 Start the Model Server container:
 
 ```bash
-docker run -d -v <folder_with_downloaded_model>:/models/face-detection/1 -p 9000:9000 openvino/model_server:latest \
+docker run -d -v <folder_with_downloaded_model>:/models/face-detection -p 9000:9000 openvino/model_server:latest \
```
Comment (pgierasi) [inline docs/ovms_quickstart.md:63]: This will not work, no models/face-detection directory

File: docs/architecture.md [hunk 9]
```diff
@@ -1,22 +1,21 @@
...
-In the backend it uses _Inference Engine libraries_ from OpenVINO&trade; toolkit, which speeds up the execution on CPU,
-and enables it on AI accelerators like NCS, iGPU, HDDL. 
+- OpenVINO&trade; Model Server is a C++ implementation of gRPC and RESTful API interfaces defined by [Tensorflow Serving](https://www.tensorflow.org/tfx/guide/serving).
 
-gRPC code skeleton is created based on TensorFlow Serving core framework with tunned implementation of requests handling. 
-Services are designed via set of C++ classes managing AI models in Intermediate Representation 
-format. OpenVINO Inference Engine component executes the graphs operations.
+- OpenVINO&trade; Model Server uses [Inference Engine](https://docs.openvinotoolkit.org/latest/openvino_docs_IE_DG_inference_engine_intro.html) libraries from OpenVINO&trade; toolkit in the backend, which speeds up the execution on CPU and enables it on AI accelerators like [Neural Compute Stick 2](https://software.intel.com/content/www/us/en/develop/hardware/neural-compute-stick.html), iGPU(Integrated Graphics Processing Unit), [HDDL](https://docs.openvinotoolkit.org/2018_R5/_docs_IE_DG_supported_plugins_HDDL.html) and FPGAs.
```
Comment (pgierasi) [inline docs/architecture.md:5]: FPGA?

File: docs/architecture.md [hunk 10]
```diff
@@ -1,22 +1,21 @@
...
+</p>
 
-The only two exposed network interfaces are gRPC and RESTful API, which currently _does not_ include authorization, 
-authentication, or data encryption. Those functions are expected to be implemented outside of the model server 
-(for example via Kubernetes* ingress or nginx forwarding proxy). 
+**<div align="center">Figure 1: Docker Container (VM or Bare Metal Host)</div>**
+
+- OpenVINO&trade; Model Server requires the models to be present in the local file system or they could be hosted remotely on object storage services. Both Google Cloud Storage and S3 compatible storage are supported. Refer to [Preparing the Models Repository](./models_repository.md) for more details.
+
+- OpenVINO&trade; Model Server is suitable for landing in Kubernetes environment. It can be also hosted on a bare metal server, virtual machine or inside a docker container. It is also suitable for landing in Kubernetes environment.
```
Comment (pgierasi) [inline docs/architecture.md:19]: suitable for landing in Kubernetes environment - mentioned twice

File: docs/ModelServerGRPCAPI.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+Sends requests via TFS gRPC API using images in numpy format. It displays performance statistics and optionally the model accuracy.
+
+[Predict proto](https://github.com/tensorflow/serving/blob/r1.14/tensorflow_serving/apis/predict.proto) has two message definitions: *PredictRequest* and  *PredictResponse*.  
+ * *PredictRequest* specifies information about the model spec, a map of input data serialized via 
+[TensorProto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.proto) to a string format.
+ * *PredictResponse* includes a map of outputs serialized by 
+[TensorProto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.proto) and information about the used model spec.
+
+There are two ways in which gRPC request can be submitted for Predict API:
```
Comment (pgierasi) [inline docs/ModelServerGRPCAPI.md:51]: This is not true

File: docs/docker_container.md [hunk 12]
```diff
@@ -1,165 +1,115 @@
-# Using the OpenVINO&trade; Model Server in a Docker Container
+# Installing OpenVINO&trade; Model Server for Linux using Docker Container
```
Comment (pgierasi) [inline docs/docker_container.md:1]: Installing using docker?

File: README.md [hunk 13]
```diff
@@ -17,23 +17,23 @@ Review the [Architecture concept](docs/architecture.md) document for more detail
 
 A few key features: 
 - Support for multiple frameworks. Serve models trained in popular formats such as Caffe*, TensorFlow*, MXNet* and ONNX*.
-- Deploy new [model versions](https://github.com/openvinotoolkit/model_server/blob/main/docs/docker_container.md#model-version-policy) without changing client code.
+- Deploy new [model versions](docs/ModelVersionPolicy.md) without changing client code.
```
Comment (pgierasi) [inline README.md:20]: without OVMS restart

File: docs/ensemble_scheduler.md [hunk 14]
```diff
@@ -1,61 +1,87 @@
...
+This guide gives information about following :
 
-## Example use case
-Let's say you want to develop an application to perform image classification. There are many different models you can use for this task. What we want to achieve is to combine results from inferences executed on two different models and calculate argmax to pick most probable classification label. For this task we select two models: [googlenet-v2](https://docs.openvinotoolkit.org/latest/omz_models_public_googlenet_v2_tf_googlenet_v2_tf.html) and [resnet-50](https://docs.openvinotoolkit.org/latest/omz_models_public_resnet_50_tf_resnet_50_tf.html). We will also create our own model **argmax** to combine and select top result. We want to perform this task on the server side with no intermediate results passed over the network. Server should take care of feeding inputs/outputs in subsequent models. Both - googlenet and resnet predictions should run in parallel. Diagram for this pipeline would look like this: 
+* <a href="#node-type">Node Types</a>
+    * Pre-defined Node Types
+    * Other Node Types
+* <a href="#example">Example Use Case</a>
+    1. Prepare the models
```
Comment (pgierasi) [inline docs/ensemble_scheduler.md:12]: i, ii, iii, iv, v?

File: docs/ensemble_scheduler.md [hunk 15]
```diff
@@ -1,61 +1,87 @@
...
+    - This node defines which inputs are required to be sent via gRPC/REST request for pipeline usage. You can refer to it by node name: `request`.
+* Response node
+    - This node defines which outputs will be fetched from final pipeline state and packed into gRPC/REST response. You cannot refer to it in your pipeline since it is pipeline final stage. To define final outputs fill `outputs` field. 
+
+> **NOTE:** Read <a href="#define-models">below</a> for example pipeline configuration.
+
+### Other node types
+Internal pipeline nodes are created by user. Currently there is only one node type that a user can create:
+* DL model
+    - This node contains underlying OpenVINO&trade; model and performs inference on selected target device. This can be defined in configuration file. Each model input needs to be mapped to some node's `data_item` - be it input from gRPC/REST request or another `DL model` output. Results of this node's inference may be mapped to another node's input or `response` node meaning it will be exposed in gRPC/REST response.
```
Comment (pgierasi) [inline docs/ensemble_scheduler.md:32]: "be it input from" ?

File: docs/ensemble_scheduler.md [hunk 16]
```diff
@@ -1,61 +1,87 @@
...
+~$ source .tf_env/bin/activate
+~$ pip3 install tensorflow==1.15
 ```
-virtualenv -p python3 .tf_env
-source .tf_env/bin/activate
-pip3 install tensorflow==1.15
-python3 tests/models/argmax_sum.py --input_size 1001 --export_dir ~/models/tf_argmax
+3. Prepare argmax model with `(1, 1001)` input shapes to match output of googlenet and resnet output shapes. Generated model will sum inputs and calculate the index with the highest value. The model output will indicate the most likely predicted class from the ImageNet* dataset. <a name="point-3"></a>
+```
+~$ python3 tests/models/argmax_sum.py --input_size 1001 --export_dir models/public/argmax/saved_model
```
Comment (pgierasi) [inline docs/ensemble_scheduler.md:62]: incorrect export direcotry, should be under ~/models/

File: docs/ModelVersionPolicy.md [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+Model version policy makes it possible to decide which versions of model will be served by OVMS. This parameter allows you to control the memory consumption of the server and decide which versions will be used regardless of what is located under the path given when the server is started. model_version_policy parameter is optional. By default server serves only the latest version for the model. Accepted format for parameter in CLI and in config is json.
+
+Accepted values :
+
+{"all": {}}
+{"latest": { "num_versions": Integer}}
+{"specific": { "versions": List }}
+Examples:
+
+{"latest": { "num_versions":2 }} # server will serve only 2 latest versions of model
```
Comment (pgierasi) [inline docs/ModelVersionPolicy.md:12]: new lines not visible in browser

File: docs/ModelVersionPolicy.md [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+{"specific": { "versions":[1, 3] }} # server will serve only 1 and 3 versions of given model
+{"all": {}} # server will serve all available versions of given model
+
+## Updating model versions
+- Served versions are updated online by monitoring file system changes in the model storage. OpenVINO Model Server will add new version to the serving list when new numerical subfolder with the model files is added. The default served version will be switched to the one with the highest number. 
+
+- When the model version is deleted from the file system, it will become unavailable on the server and it will release RAM allocation. Updates in the deployed model version files will not be detected and they will not trigger changes in serving.
```
Comment (pgierasi) [inline docs/ModelVersionPolicy.md:19]: Updates in the deployed deleted model?

File: docs/AgeGenderGuide.md [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+data_json = json.dumps(data_obj)
+
+result = requests.post("http://localhost:9001/v1/models/age_gender:predict", data=data_json)
+result_dict = json.loads(result.text)
+print(result_dict)
+```
+
+3. Run the above code snippet to send POST API request to predict results by providing formatted json as request body using the command :
+```Bash
+python3 sample.py
```
Comment (pgierasi) [inline docs/AgeGenderGuide.md:69]: example not working

File: docs/FaceDetectionScriptExample.md [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+                        Specify the model name 
+```
+
+### Usage example
+
+Start the OVMS service locally:
+
+```bash
+curl --create-dirs https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/1/face-detection-retail-0004/FP32/face-detection-retail-0004.xml https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/1/face-detection-retail-0004/FP32/face-detection-retail-0004.bin -o model/1/face-detection-retail-0004.xml -o model/1/face-detection-retail-0004.bin
+docker run -d -v $(pwd)/model:/models -p 9000:9000 openvino/model_server:latest  --model_path /models --model_name face-detection --port 9000  --shape auto
```
Comment (pgierasi) [inline docs/FaceDetectionScriptExample.md:47]: [2020-10-22 21:47:38.396] [serving] [error] Couldn't list directories in path: /models
[2020-10-22 21:47:38.396] [serving] [error] Couldn't start model manager
[2020-10-22 21:47:38.396] [serving] [error] ovms::ModelManager::Start() Error: The provided base path is invalid or doesn't exists

File: docs/host.md [hunk 21]
```diff
@@ -1,94 +1,118 @@
...
-You can build it inside a docker container. The building steps are automated 
-with a command:
-```bash
-make docker_build
+> **NOTES**:
+> * These steps apply to Ubuntu*, CentOS* and RedHat*
+> * An internet connection is required to follow the steps in this guide.
+
+## Introduction
+OpenVINO&trade; Model Server is a Python* implementation of gRPC and RESTful API interfaces defined by Tensorflow serving.
```
Comment (pgierasi) [inline docs/host.md:8]: is a Python implementation?

**Review by mzegla — APPROVED** [review]
From my side it's okay, but I see you didn't make all the changes requested by @pgierasi in example_client/README.md

**Comment by dtrawins** [comment]
@mzegla @pgierasi all requested changes in example_client/README.md are already added

**Review by pgierasi — APPROVED** [review]

**Review by pgierasi — APPROVED** [review]

**Review by pgierasi — APPROVED** [review]
